### ISSUE TITLE
[dls-algorithm-for-formula-equations] false solution on ∃X (X(a) ∧ ¬X(b))

### ISSUE THREAD

**Issue by shetzl** [gapt/gapt#745]
Doing the following in the CLI
```
import gapt.logic.hol._
val F = hof"∃X (X(a) ∧ ¬X(b))"
val solution = solveFormulaEquation( F )
```
results in
```
solution: scala.util.Try[(gapt.expr.subst.Substitution, gapt.expr.formula.Formula)] = Success((Substitution(X:i>o -> λx_0 a = b ∨ a = x_0),X(a) ∧ ¬X(b)))
```
which is not a solution for the formula equation F and hence
```
Escargot.isValid(BetaReduction.betaNormalize(solution.get._1(solution.get._2)))
```
returns `false`.

**Comment by fachammer** [comment]
hm, but isn't it actually a solution? Using that X(a) is equivalent to ∀x (x=a -> X(x)) and applying Ackermann's Lemma gives ∃X (X(a) ∧ ¬X(b)) <-> (¬X(b))[X / λx x=a] <-> ¬(a=b).
Substituting the witness which is returned by `solveFormulaEquation` also simplifies to ¬(a=b). In that sense substituting the witness gives a first-order formula which is equivalent to the original formula equation, which is what we want, right?

I'm not sure why Escargot returns `false` in this case, but maybe it has something to do with constants still appearing in the first-order formula?

Also, I see that in this case the simpler witness λx_0 x_0=a could be returned. I will push a fix shortly.

**Comment by jvierling** [comment]
@shetzl What is the expected outcome of the example you gave? As far as I see the formula equation `X(a) ∧ ¬X(b)` does not have any solutions, because there are structures (of the theory ∅) with `a = b`.

**Comment by jvierling** [comment]
@fachammer 

- A solution of a formula equation `F[X]` is a formula `G` such that F[G] is valid i.e. we do not care about the equivalence `∃XF[X] ↔ F[G]`. In our setting we can use the "DLS" algorithm to solve formula equations if `∃XF[X]` is valid, which is however not the case of the formula in the example. 
- Escargot returns `false` because the formula `BetaReduction.betaNormalize(solution.get._1(solution.get._2))` (i.e. `(a = a ∨ a ≠ b) ∧ (a = b ∨ a = b)`) is not valid.

**Comment by shetzl** [comment]
Yes, indeed, this formula equation is not solvable. The expected outcome is a failure of solveFormulaEquation.

**Comment by fachammer** [comment]
@shetzl @jvierling Ah, I see. Then I've misunderstood what it means to solve a formula equation. How do we deal with that though, since the original DLS algorithm wouldn't fail in this case either? I guess we could run `Escargot.isValid` inside of `solveFormulaEquation` and return `Failure` when we encounter first-order formula which is not valid. Is that what you'd expect?

**Comment by jvierling** [comment]
@fachammer I think we should just rename `solveFormulaEquation` to some name that expresses that the function actually does a quantifier-elimination with witness extraction. We could then add another function that solves formula equations, by using the quantifier-elimination followed by an `Escargot.isValid` check.

**Comment by fachammer** [comment]
@jvierling @shetzl I've made the changes as you suggested now with a method `witnessForSecondOrderQuantifierElimination` which does what `solveFormulaEquation` used to do. The new `solveFormulaEquation` now checks for validity before returning and only returns the substitution, without the first order part. Also, I've updated the user manual to reflect these changes.

**Comment by jvierling** [comment]
Fixed by a74634f
